runtime/main/src/main/java/org/qbicc/runtime/main/VMHelpers.java

```diff
@@ -175,16 +175,206 @@ static void raiseHeapDeserializationError() {
     }
 
     // TODO: mark this with a "NoInline" annotation
-    static void raiseIncompatibleClassChangeError() { throw new IncompatibleClassChangeError(); }
+    static void raiseIncompatibleClassChangeError() {
+        throw new IncompatibleClassChangeError(); 
+    }
 
     // TODO: mark this with a "NoInline" annotation
-    static void raiseNegativeArraySizeException() { throw new NegativeArraySizeException(); }
+    static void raiseNegativeArraySizeException() {
+        throw new NegativeArraySizeException(); 
+    }
 
     // TODO: mark this with a "NoInline" annotation
     static void raiseNullPointerException() {
         throw new NullPointerException();
     }
 
     // TODO: mark this with a "NoInline" annotation
-    static void raiseUnsatisfiedLinkError() { throw new UnsatisfiedLinkError(); }
-}
+    static void raiseUnsatisfiedLinkError() {
+        throw new UnsatisfiedLinkError(); 
+    }
+ 
+    // TODO: mark with "must be build time initialized" annotation
+    static class ClinitState {
+        static final byte STATE_UNINITIALIZED = 0;
+        static final byte STATE_INPROGRESS = 1;
+        static final byte STATE_INITIALIZED = 2;
+        static final byte STATE_FAILED = 3;
+
+        Thread initializerThread;
+        Throwable errorInInitializer;
+        // Must be one of:
+        // 0 - uninitialized
+        // 1 - in-progress
+        // 2 - initialized
+        // 3 - failed
+        byte initializedState;
```

Comment:
There are cases were we might have a `typeid` - and need initialization - but won't need the `Class` object to be present in the system.  

We had previously discussed not having a `Class` for every "class" in the system and at the time, I wasn't sold on the idea, but I'm coming around as there will be classes were we use a static member and need the proper initialization, but never need to materialize the `Class` object.

If we move these to the `Class`, we'll need a class object for any class that has a static initializer.

Reply:
Well now I'm not so sure. :-)

In order to skip a `Class` object, we'd have to prove that `getClass()` is never called on instances of that (concrete) class, or on instances of any of its subclasses, *or* on references whose lower bound might include it (i.e. if we end up with even one call to `Object#getClass()` at run time, all classes must have `Class` objects visible at run time).  Additionally, we cannot skip `Class` objects for any supertypes of any `Class` object we realize (unless we can prove that `Class.getSuperclass()` and `Class.getInterfaces()` is never called for a given class), so that rules out more potentially wide space.  `Class` objects must be present for any objects whose reflection data is present in the image so we can return them for field types and method parameter and return types, among other things.  Also, any class with a method that can appear in a call stack must have `Class` unless we can prove that `java.lang.StackWalker.StackFrame#getDeclaringClass` is never called, so that basically rules out all classes with methods.  That is likely to leave us with just empty classes, which will be deleted at run time, and non-instantiable non-constant-static-field-only classes, of which there must be very few.  That's just off the top of my head.

Furthermore, if we use compact scheme for storing `Class` objects (specifically, an array of `Class` objects such that a reference to a given `Class` object can be computed with zero loads by adding the `type_id` to the base of the `Class` array), we'd still have to reserve the same space for the `Class` objects anyway.

So it seems to me that it's probably not worthwhile to try to reduce the number of `Class` objects below the number of actual classes present at run time.  It seems like a lot of effort for probably little return.

Reply:
> So it seems to me that it's probably not worthwhile to try to reduce the number of Class objects below the number of actual classes present at run time. It seems like a lot of effort for probably little return.

I agree with this point for normal classes.  I assume we're still going to dynamically create Class instances for specific reference array types on demand though.

Reply:
If we're committing to having Class objects, then the initializeState should definitely migrate to the `j.l.Class`

Reply:
Added a `/* TODO */` comment in https://github.com/qbicc/qbicc/pull/530/ to indicate this state should likely move to the Class object once we have them